package com.alhuda.app.core.presentation.navigation

import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.SnackbarHostState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.CompositionLocalProvider
import androidx.compose.runtime.remember
import androidx.lifecycle.viewmodel.navigation3.rememberViewModelStoreNavEntryDecorator
import androidx.navigation3.runtime.NavKey
import androidx.navigation3.runtime.entryProvider
import androidx.navigation3.runtime.rememberNavBackStack
import androidx.navigation3.runtime.rememberSaveableStateHolderNavEntryDecorator
import androidx.navigation3.ui.NavDisplay
import androidx.savedstate.serialization.SavedStateConfiguration
import com.alhuda.app.core.presentation.LightColorScheme
import com.alhuda.app.core.presentation.components.LocalSnackbarController
import com.alhuda.app.core.presentation.components.SnackbarController
import com.alhuda.app.core.presentation.feedback.ObserveScheduleFeedback
import com.alhuda.app.intro.IntroNavigation
import com.alhuda.app.main.MainNavigation
import kotlinx.serialization.modules.SerializersModule
import kotlinx.serialization.modules.polymorphic

@Composable
fun NavigationRoot(
    appIntroDone: Boolean,
    startingRoute: Route?,
) {
    val rootBackStack =
        rememberNavBackStack(
            configuration =
                SavedStateConfiguration {
                    serializersModule = SerializersModule {
                        polymorphic(NavKey::class) {
                            subclass(Route.Intro::class, Route.Intro.serializer())
                            subclass(Route.Main::class, Route.Main.serializer())
                        }
                    }
                },
            if (appIntroDone) Route.Main else Route.Intro,
        )

    CompositionLocalProvider(
        LocalSnackbarController provides remember { SnackbarController(SnackbarHostState()) },
    ) {
        ObserveScheduleFeedback()
        NavDisplay(
            backStack = rootBackStack,
            entryDecorators =
                listOf(
                    rememberSaveableStateHolderNavEntryDecorator(),
                    rememberViewModelStoreNavEntryDecorator(),
                ),
            entryProvider =
                entryProvider {
                    entry<Route.Intro> {
                        MaterialTheme(colorScheme = LightColorScheme) {
                            IntroNavigation(
                                onFinishIntro = {
                                    rootBackStack.clear()
                                    rootBackStack.add(Route.Main)
                                },
                            )
                        }
                    }

                    entry<Route.Main> {
                        MainNavigation(
                            startingRoute,
                        )
                    }
                },
        )
    }
}
